### PR TITLE
Part 2: Make enum -> enum class

### PR DESCRIPTION
**Pull request by TobiKattmann**
Hi all, 

this is the continuation of #1242 . Change of the `enum`'s in `option_structure.hpp` into `enum class`. This has the benefit of no implicit conversion to int and therefore increased safety with the enum type -> [c++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum3-prefer-class-enums-over-plain-enums).

If you want to sponsor a commit (which I highly recommend ;) ) then a good approach is to 
1. make your `enum` of choice an `enum class` and remove the associated numbers. 
2. Rename here if reasonable, i,.e. `ENUM_INLET_TYPE` can be just `INLET_TYPE`. Also the option names to not be redundant so e.g. `INLET_TYPE::PRESSURE_INLET_TYPE` could become  `INLET_TYPE::PRESSURE`
3. Add a `NONE` or `NO_THING` option if needed. Often the global `NONE` option is and that made a lot of people very angry and has been widely regarded as a bad move 🧙 
2. then you go to `CConfig.c/hpp` and change the option type and everything there (because in the `CConfig` class the variable is usually defined). Find the `GetKind_yourOption` and search the code base for that `Get` function.
3. Once all search results and the following function bodies are changed just compile and fix until the code builds
4. Special care for switch statements: If not all options are present make a `default; break;` line -> otherwise the automated builds will fail on a warning (clang is used, I found gcc does only complain with an additional flag)
4. commit at best just 1 commit with a clear message and push. That makes reverting easier if it fails and introduced some non-obvious error
 
Have fun :)

## Related Work
#1242

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [ ] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [ ] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: Common/include/option_structure.hpp [hunk 1]
```diff
@@ -2070,11 +2070,12 @@ static const MapType<string, ENUM_DIRECTDIFF_VAR> DirectDiff_Var_Map = {
 };
 
 
-enum ENUM_RECORDING {
-  SOLUTION_VARIABLES = 1,
-  MESH_COORDS = 2,
-  MESH_DEFORM = 3,
-  SOLUTION_AND_MESH = 4
+enum class RECORDING {
+  NO_RECORDING,
```
Comment (TobiKattmann) [inline Common/include/option_structure.hpp:2074]: Added a `NO_RECORDING` option as the global `NONE` was used

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp [hunk 2]
```diff
@@ -235,9 +235,9 @@ void CDiscAdjMultizoneDriver::Run() {
 
   /*--- Evaluate the objective function gradient w.r.t. the solutions of all zones. ---*/
 
-  SetRecording(NONE, Kind_Tape::OBJECTIVE_FUNCTION_TAPE, ZONE_0);
-  SetRecording(SOLUTION_VARIABLES, Kind_Tape::OBJECTIVE_FUNCTION_TAPE, ZONE_0);
-  RecordingState = NONE;
+  SetRecording(RECORDING::NO_RECORDING, Kind_Tape::OBJECTIVE_FUNCTION_TAPE, ZONE_0);
```
Comment (TobiKattmann) [inline SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: see here

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: We can try to make the "NONE" case either expressive or concise, for example the NONE of time marching is steady.
The NONE recording serves to clear indices.

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: We can try to make the "NONE" case either expressive or concise, for example the NONE of time marching is steady.
The NONE recording serves to clear indices.

Comment (TobiKattmann) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: Exactly. For TIME_MARCHING that is done 👍 
Only add a NONE-type option if it not already expressed by another option by an elimination process

Comment (TobiKattmann) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: oh and 👍 for the addition :)

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: What I'm say is RECORDING::NO_RECORDING is not a very good name IMO

Comment (TobiKattmann) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: I'll change it to RECORDING::CLEAR_INDICES (tomorrow). That should be better?
(Sorry I read the comment with a vanished attention span 😴 :) )

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:238]: I think so, gives more info at least, but hey... I'm sleeping on the job too.

File: Common/include/option_structure.hpp [hunk 3]
```diff
@@ -469,20 +469,6 @@ const int EL_PYRAM = 2;   /*!< \brief Elements of five nodes (3D). */
 const int EL_PRISM = 3;   /*!< \brief Elements of six nodes (3D). */
 
 
-/*!
- * \brief Types of mathematical problem to solve
- */
-enum ENUM_MATH_PROBLEM {
```
Comment (TobiKattmann) [inline Common/include/option_structure.hpp:503]: not used any more, nice 👍

**Review by pcarruscag — APPROVED** [review]

**Comment by TobiKattmann** [comment]
Codefactor is just complainig about the usual CIncEulerSolver. All other test pass, so I'll merge this and continue in Part III. 
Coming to a repository near you soon. 🎦

**Pull request merged**